cedar-lean/Cedar/Thm/Lemmas/Typechecker.lean

```diff
@@ -0,0 +1,175 @@
+  ∀ uid data, entities.find? uid = some data →
+    ∃ attrTys ancestorTys, ets.find? uid.ty = some (attrTys, ancestorTys) ∧
+      InstanceOfType data.attrs (.record attrTys) ∧
+      ∀ ancestor, ancestor ∈ data.ancestors → ancestor.ty ∈ ancestorTys
+
+/--
+For every action in the entity store, the action's ancestors are consistent
+with the ancestor information in the action store.
+-/
+def InstanceOfActionStore (entities : Entities) (as: ActionStore) : Prop :=
+  ∀ uid data, entities.find? uid = some data →
+    ∃ ancestors, as.find? uid = some ancestors →
+      ∀ ancestor, ancestor ∈ data.ancestors → ancestor ∈ ancestors
+
+def RequestAndEntitiesMatchEnvironment (env : Environment) (request : Request) (entities : Entities) : Prop :=
+  InstanceOfRequestType request env.reqty ∧
+  InstanceOfEntityTypeStore entities env.ets ∧
+  InstanceOfActionStore entities env.acts
+
+/--
+The type soundness property says that if the typechecker assigns a type to an
+expression, then it must be the case that the expression `EvaluatesTo` a value
+of that type. The `EvaluatesTo` predicate covers the (obvious) case where evaluation
+has no errors, but it also allows for errors of type `entityDoesNotExist` and
+`extensionError`.
+
+The typechecker cannot protect against these errors because they depend on runtime
+information (i.e., the entities passed into the authorization request, and
+extension function applications on authorization-time data). All other errors
+(`attrDoesNotExist` and `typeError`) can be prevented statically.
+
+_Note_: Currently, `extensionError`s can also be ruled out at validation time
+because the only extension functions that can error are constructors, and all
+constructors are required to be applied to string literals, meaning that they
+can be fully evaluated during validation. This is not guaranteed to be the case
+in the future.
+-/
+def EvaluatesTo (e: Expr) (request : Request) (entities : Entities) (v : Value) : Prop :=
+  evaluate e request entities = .error .entityDoesNotExist ∨
+  evaluate e request entities = .error .extensionError ∨
+  evaluate e request entities = .ok v
+
+/--
+On input to the typechecking function, for any (e,k) in the Capabilities,
+e is a record- or entity-typed expression that has key k.
+-/
+def CapabilitiesInvariant (c : Capabilities) (request : Request) (entities : Entities) : Prop :=
+  ∀ (e : Expr) (k : Attr), (e, k) ∈ c → EvaluatesTo (.hasAttr e k) request entities true
+
+/--
+The Capabilities output by the typechecking function will satisfy
+`CapabilitiesInvariant` provided that the input expression evaluates to true.
+-/
+def GuardedCapabilitiesInvariant (e: Expr) (c: Capabilities) (request : Request) (entities : Entities) : Prop :=
+  evaluate e request entities = .ok true →
+  CapabilitiesInvariant c request entities
+
+-- Easy property: the empty capability set satisifies the invariant
+theorem empty_CapabilitiesInvariant (request : Request) (entities : Entities) :
```

Comment:
I personally don't like this mixed naming style

Reply:
Yeah it's a little ugly. Switched to a consistent style for lemma names in the most recent commit.